src/ripple/ledger/tests/View_test.cpp

```diff
@@ -524,6 +525,209 @@ class View_test
+            auto const gwUSD = accountFunds(
+                *env.closed(), gw, USD(0), fhZERO_IF_FROZEN, env.journal);
+            expect (gwUSD == USD(0));
+
+            // carol has funds from the gateway.
+            auto carolsUSD = accountFunds(
+                *env.closed(), carol, USD(0), fhZERO_IF_FROZEN, env.journal);
+            expect (carolsUSD == USD(50));
+
+            // If carols funds are frozen she has no funds...
+            env(fset (gw, asfGlobalFreeze));
+            env.close();
+            carolsUSD = accountFunds(
+                *env.closed(), carol, USD(0), fhZERO_IF_FROZEN, env.journal);
+            expect (carolsUSD == USD(0));
+
+            // ... unless the query ignores the FROZEN state.
+            carolsUSD = accountFunds(
+                *env.closed(), carol, USD(0), fhIGNORE_FREEZE, env.journal);
+            expect (carolsUSD == USD(50));
+
+            // Just to be tidy, thaw gw.
+            env(fclear (gw, asfGlobalFreeze));
+            env.close();
+        }
+    }
+
+    void
+    testTransferRate()
+    {
+        using namespace jtx;
+        Env env(*this);
+
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+        auto const gw1 = Account("gw1");
+
+        env.fund(XRP(10000), alice, bob, gw1);
+        env.close();
+
+        auto rdView = env.closed();
+        // Test with no rate set on gw1.
+        expect (rippleTransferRate (*rdView, alice, bob, gw1) == 1000000000);
+        expect (rippleTransferRate (*rdView, gw1, alice, gw1) == 1000000000);
+        expect (rippleTransferRate (*rdView, alice, gw1, gw1) == 1000000000);
+
+        env(rate(gw1, 1.02));
+        env.close();
+
+        rdView = env.closed();
+        // Test with a non-unity rate set on gw1.
+        expect (rippleTransferRate (*rdView, alice, bob, gw1) == 1020000000);
+        expect (rippleTransferRate (*rdView, gw1, alice, gw1) == 1000000000);
+        expect (rippleTransferRate (*rdView, alice, gw1, gw1) == 1000000000);
+    }
+
+    void
+    testAreCompatible()
+    {
+        // This test requires incompatible ledgers.  The good news is that
+        // ledgers have their memory managed by shared_ptr.  So we can:
+        //  o Make an Env and construct a couple of ledgers in it,
+        //  o Destroy the Env that created them, but save the ledgers,
+        //  o Make a new Env, and
+        //  o Use the new Env to make a couple of ledgers that are
+        //    incompatible with the first two.
+        using namespace jtx;
+        auto const alice = Account("alice");
+        auto const bob = Account("bob");
+
+        // Make a couple of ledgers, then destroy their environment.
+        std::shared_ptr<const ReadView> rdViewA3;
+        std::shared_ptr<const ReadView> rdViewA4;
+        {
+            Env env(*this);
+
+            env.fund(XRP(10000), alice);
+            env.close();
+            rdViewA3 = env.closed();
```

Comment:
We should probably audit the code to make sure that ledgers can outlive the `Application` object. Or if not, figure out what the obstacles are. I think `SHAMapFamily` could be one road block.